File: lib/mercuryapi-1.23.0.20/java/samples_nb/src/samples/RebootReader.java
/*
 * Sample program that reboots the reader
 * 
 */
package samples;

import com.thingmagic.*;


public class RebootReader 
{
      
  static void usage()
  {
    System.out.printf("Usage: demo reader-uri <command> [args]\n" +
                      "  (URI: 'tmr:///COM1' or 'tmr://astra-2100d3/' " +
                      "or 'tmr:///dev/ttyS0')\n\n" +
                      "Available commands:\n");
    System.exit(1);
  }

   public static void setTrace(Reader r, String args[])
  {    
    if (args[0].toLowerCase().equals("on"))
    {
      r.addTransportListener(r.simpleTransportListener);
    }    
  }
  
  public static void main(String argv[])
  {
    // Program setup
    Reader r = null;
    int nextarg = 0;
    boolean trace = false;

    if (argv.length < 1)
      usage();

    if (argv[nextarg].equals("-v"))
    {
      trace = true;
      nextarg++;
    }
    
    // Create Reader object, connecting to physical device
    try
    {
      
      r = Reader.create(argv[nextarg]);
      if (trace)
      {
        setTrace(r, new String[] {"on"});
      }
      r.connect();
      if (Reader.Region.UNSPEC == (Reader.Region)r.paramGet("/reader/region/id"))
      {
          Reader.Region[] supportedRegions = (Reader.Region[])r.paramGet(TMConstants.TMR_PARAM_REGION_SUPPORTEDREGIONS);
          if (supportedRegions.length < 1)
          {
               throw new Exception("Reader doesn't support any regions");
      }
          else
          {
               r.paramSet("/reader/region/id", supportedRegions[0]);
          }
      }
      r.reboot();      
      r.destroy();
      boolean notConnected = true;
      int retryCount =1;
      do
      {
          try
          {
              System.out.println("Trying to reconnect.... Attempt:"+retryCount);
              r = Reader.create(argv[nextarg]);
              r.connect();
              notConnected = false;
          }
          catch(Exception ex)
          {
          }
          retryCount++; 
      }while(notConnected);
    } 
    catch (ReaderException re)
    {
      System.out.println("Reader Exception : " + re.getMessage());
    }
    catch (Exception re)
    {
        System.out.println("Exception : " + re.getMessage());
    }
  }
}

File: lib/mercuryapi-1.23.0.20/cs/Samples/Codelets/RebootReader/RebootReader.cs
using System;
using System.Collections.Generic;
using System.Text;

// Reference the API
using ThingMagic;

namespace RebootReader
{
    /// <summary>
    /// Sample program that reboots the reader
    /// </summary>
    class Program
    {
        static void Main(string[] args)
        {
            // Program setup
            if (1 != args.Length)
            {
                Console.WriteLine(String.Join("\r\n", new string[] {
                    "Please provide reader URL, such as:",
                    "tmr:///com4",
                    "tmr://my-reader.example.com",
                }));
                Environment.Exit(1);
            }

            try
            {
                // Create Reader object, connecting to physical device.
                // Wrap reader in a "using" block to get automatic
                // reader shutdown (using IDisposable interface).
                Reader r;
                using (r = Reader.Create(args[0]))
                {
                    //Uncomment this line to add default transport listener.
                    //r.Transport += r.SimpleTransportListener;

                    r.Connect();
                    if (Reader.Region.UNSPEC == (Reader.Region)r.ParamGet("/reader/region/id"))
                    {
                        Reader.Region[] supportedRegions = (Reader.Region[])r.ParamGet("/reader/region/supportedRegions");
                        if (supportedRegions.Length < 1)
                        {
                            throw new FAULT_INVALID_REGION_Exception();
                        }
                        else
                        {
                            r.ParamSet("/reader/region/id", supportedRegions[0]);
                        }
                    }
                    r.Reboot();
                    r.Destroy();
                    int retryCount = 1;
                    bool notConnected = true;

                    /*For universal reader URI scheme i.e tmr: the create has to be inside the loop. 
                     * In order to use the create method out side the loop use product specific URI
                     * scheme i.e llrp: or rql:                     
                     */
                    do
                    {
                        try
                        {
                            Console.WriteLine("Trying to reconnect.... Attempt: " + retryCount);
                            r = Reader.Create(args[0]);
                            r.Connect();
                            notConnected = false;
                        }
                        catch (Exception)
                        {
                            
                        }
                        retryCount++;
                    } while (notConnected);
                    Console.WriteLine("Reader is reconnected successfully");
                }
            }
            catch (ReaderException re)
            {
                Console.WriteLine("Error: " + re.Message);
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error: " + ex.Message);
            }
        }
    }
}
